contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -639,8 +639,29 @@ CAPIStatus Filter::setHeader(absl::string_view key, absl::string_view value) {
     ENVOY_LOG(debug, "invoking cgo api at invalid phase: {}", __func__);
     return CAPIStatus::CAPIInvalidPhase;
   }
-  headers_->setCopy(Http::LowerCaseString(key), value);
-  onHeadersModified();
+  if (state.isThreadSafe()) {
+    // it's safe to write header in the safe thread.
+    headers_->setCopy(Http::LowerCaseString(key), value);
+    onHeadersModified();
+  } else {
```

Comment:
It's needs to add comments to explain that the API does not take effect immediately, and the API on the GO side comments also.

Reply:
Done, thanks.
Also, changed prefetch the values in `header.Set` in the Go side, so that users won't get outdated values, in the following case:
```
# async goroutine, in Go thread
header.Set("foo", "bar")
header.Get("foo")
```